src/metabase/models/field_usage.clj

```diff
@@ -0,0 +1,100 @@
+(ns metabase.models.field-usage
+  (:require
+   [metabase.lib.core :as lib]
+   [metabase.lib.dispatch :as lib.dispatch]
+   [metabase.lib.hierarchy :as lib.hierarchy]
+   [metabase.lib.schema :as lib.schema]
+   [metabase.lib.util :as lib.util]
+   [metabase.lib.util.match :as lib.util.match]
+   [metabase.models.interface :as mi]
+   [metabase.query-processor.middleware.fetch-source-query :as fetch-source-query]
+   [metabase.util.malli :as mu]
+   [methodical.core :as methodical]
+   [toucan2.core :as t2]
+   [toucan2.tools.disallow :as t2.disallow]))
+
+(doto :model/FieldUsage
+  (derive :metabase/model)
+  (derive ::t2.disallow/update)
+  (derive :hook/created-at-timestamped?))
+
+(methodical/defmethod t2/table-name :model/FieldUsage [_model] :field_usage)
+
+(t2/deftransforms :model/FieldUsage
+  {:used_in                mi/transform-keyword
+   :aggregation_function   mi/transform-keyword
+   :breakout_temporal_unit mi/transform-keyword
+   :breakout_binning       mi/transform-json
+   :filter_op              mi/transform-keyword
+   :filter_args            mi/transform-json})
+
+(defn- filter->field-usage
+  [query stage filter-clause]
+  (let [filter-parts (lib/filter-parts query stage filter-clause)
+        field-id     (-> filter-parts :column :id)]
+    (when (int? field-id)
+      {:field_id    field-id
+       :used_in     :filter
+       :filter_op   (-> filter-parts :operator :short)
+       :filter_args (:args filter-parts)})))
+
+(defn- aggregation->field-usage
+  [query stage aggregation-clause]
+  (let [aggregation-column (lib/aggregation-column query stage aggregation-clause)
+        field-id           (:id aggregation-column)]
+    (when (int? field-id)
+      {:field_id             field-id
+       :used_in              :aggregation
+       :aggregation_function (first aggregation-clause)})))
+
+(defn- breakout->field-usage
+  [query stage breakout-clause]
+  (let [breakout-column (lib/breakout-column query stage breakout-clause)
+        field-id        (:id breakout-column)]
+    (when (int? field-id)
+      {:field_id               field-id
+       :used_in                :breakout
+       :breakout_temporal_unit (lib/raw-temporal-bucket breakout-clause)
+       :breakout_binning       (-> (lib/binning breakout-clause)
+                                   (select-keys [:strategy :bin-width :num-bins])
+                                   not-empty)})))
```

Comment:
yea that's a reasonable desire. but otoh this is going to be a big table and has too many nullable columns will waste a lot of space. not sure if it's worth it.

Reply:
If you're worried about space in the DB (and I'm not sure we are), breaking it into multiple columns should actually use less space, not more. It seems counterintuitive but the reason is because you need to have a wide column to store JSON data, which dominates any overhead of having the extra columns.

I was curious and tested this with postgres just to be sure:

```
testdb=# -- Create the first table
CREATE TABLE table1 (
  id SERIAL PRIMARY KEY,
  breakout_binning VARCHAR(256)
);

-- Insert a dummy row with example values into table1
INSERT INTO table1 (breakout_binning)
VALUES ('Example binning');

-- Insert 1000 rows with only the ID column having not-null values into table1
INSERT INTO table1 (breakout_binning)
SELECT NULL
FROM generate_series(1, 1000);
CREATE TABLE
INSERT 0 1
INSERT 0 1000

testdb=# select oid, relname 
from pg_class limit 1;
  oid   |              relname               
--------+------------------------------------
 635346 | test_20231201143858__mb_row_id_seq
(1 row)


testdb=# -- Create the second table
CREATE TABLE table2 (
  id SERIAL PRIMARY KEY,
  strategy VARCHAR(25),
  "bin-width" INTEGER,
  "num-bins" INTEGER
);

-- Insert a dummy row with example values into table2
INSERT INTO table2 (strategy, "bin-width", "num-bins")
VALUES ('Example strategy', 10, 5);

-- Insert 1000 rows with only the ID column having not-null values into table2
INSERT INTO table2 (strategy, "bin-width", "num-bins")
SELECT NULL, NULL, NULL
FROM generate_series(1, 1000);
CREATE TABLE
INSERT 0 1
INSERT 0 1000


testdb=# select pg_table_size(oid) from pg_class where relname = 'table1' limit 1;
 pg_table_size 
---------------
         73728
(1 row)


testdb=# select pg_table_size(oid) from pg_class where relname = 'table2' limit 1;
 pg_table_size 
---------------
         65536
```

table2 is smaller, even though it has more columns.